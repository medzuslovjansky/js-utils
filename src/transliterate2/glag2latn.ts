import { createReplacer } from './createReplacer';
import { Glagolitic } from '../constants';
import { replaceByWord } from './replaceByWord';

const punctuate = createReplacer({
  ';': '?',
  '։': '.',
  '·': ',',
  '⁖': '(',
  '჻': ')',
  '⸪': '"',
  '··': '!',
  '⸬': ':',
  '⁘': ';',
  '···': '...',
});

const softener = createReplacer({
  dь: 'ď',
  lь: 'ľ',
  nь: 'ń',
  rь: 'ŕ',
  sь: 'ś',
  tь: 'ť',
  zь: 'ź',
  ьr: 'ŕ',
  ȯr: 'r',
});

const yerify = (word: string) =>
  word.replace(/([dlnrstz])ь([cčklnr])/, '$1ė$2');

const deglagolify = createReplacer({
  [Glagolitic.Azu]: 'a',
  [Glagolitic.Otu]: 'å',
  [Glagolitic.Buky]: 'b',
  [Glagolitic.Tsi]: 'c',
  [Glagolitic.Chrivi]: 'č',
  [Glagolitic.Chrivi + Glagolitic.Yeri]: 'ć',
  [Glagolitic.Shta]: 'ć',
  [Glagolitic.Dobro]: 'd',
  [Glagolitic.Djervi]: 'đ',
  [Glagolitic.Dzelo]: 'ѕ',
  [Glagolitic.Yestu]: 'e',
  [Glagolitic.Yati]: 'ě',
  [Glagolitic.Small_Yus]: 'ę',
  [Glagolitic.Yeri]: 'ь',
  [Glagolitic.Fritu]: 'f',
  [Glagolitic.Glagoli]: 'g',
  [Glagolitic.Heru]: 'h',
  [Glagolitic.Spidery_Ha]: 'h',
  [Glagolitic.I]: 'i',
  [Glagolitic.Izhe]: 'j',
  [Glagolitic.Initial_Izhe]: 'j',
  [Glagolitic.Trokutasti_A]: 'ja',
  [Glagolitic.Iotated_Small_Yus]: 'ję',
  [Glagolitic.Yo]: 'jo',
  [Glagolitic.Yu]: 'ju',
  [Glagolitic.Iotated_Big_Yus]: 'jų',
  [Glagolitic.Kako]: 'k',
  [Glagolitic.Ljudije]: 'l',
  [Glagolitic.Myslite]: 'm',
  [Glagolitic.Latinate_Myslite]: 'm',
  [Glagolitic.Nashi]: 'n',
  [Glagolitic.Onu]: 'o',
  [Glagolitic.Yeru]: 'ȯ',
  [Glagolitic.Pokoji]: 'p',
  [Glagolitic.Ritsi]: 'r',
  [Glagolitic.Slovo]: 's',
  [Glagolitic.Sha]: 'š',
  [Glagolitic.Tvrido]: 't',
  [Glagolitic.Fita]: 'θ',
  [Glagolitic.Uku]: 'u',
  [Glagolitic.Big_Yus]: 'ų',
  [Glagolitic.Izhitsa]: 'ù',
  [Glagolitic.Vedi]: 'v',
  [Glagolitic.Yeri + Glagolitic.Izhe]: 'y',
  [Glagolitic.Yeru + Glagolitic.Izhe]: 'y',
  [Glagolitic.Zemlja]: 'z',
  [Glagolitic.Zhivete]: 'ž',
});

function glag2latn4word(word: string): string {
  return softener(yerify(deglagolify(word)));
}

export function glag2latn(text: string): string {
  return punctuate(replaceByWord(text, glag2latn4word));
}
