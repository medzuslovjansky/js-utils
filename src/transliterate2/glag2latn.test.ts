import { glag2latn } from './glag2latn';

describe('Glag → Latn', () => {
  const SANITY_1 = `\
Ⱀⰰ ⰲⱏⰸⰲⱏⰹⱎⰵⱀⱁⱄⱅⰻ ⱁⰲⱌⰰ· ⰽⱅⱁⱃⰰ ⱀⰵ ⰻⰿⱑⰾⰰ ⰲⱏⰾⱀⱘ· ⱆⰲⰻⰴⱑⰾⰰ ⰽⱁⱀⱐⰵⰲ։ Ⱂⱐⱃⰲⱏⰹ ⱅⱔⰳⰰⰾ ⱅⱔⰶⰽⱏⰹ ⰲⱁⰸ· ⰲⱅⱁⱃⱏⰹ ⱀⱁⱄⰻⰾ ⰲⰵⰾⰻⰽⱁ ⰱⱃⱑⰿⱔ· ⱅⱃⰵⱅⰺⰻ ⰱⱏⱃⰸⱁ ⰲⱁⰸⰻⰾ ⰿⱘⰶⰰ։
Ⱁⰲⱌⰰ ⱃⱑⰽⰾⰰ ⰽⱁⱀⱐⰰⰿ⸬ ⸪Ⰱⱁⰾⰻ ⰿⱀⱑ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰼⱘ· ⰽⰰⰽⱁ ⱍⰾⱁⰲⱑⰽ ⰲⰾⰰⰴⰰⰹⰵ ⰽⱁⱀⱐⰰⰿⰻ։⸪
Ⰽⱁⱀⱐⰻ ⱃⱑⰽⰾⰻ⸬ ⸪Ⱄⰾⱆⱎⰰⰹ· ⱁⰲⱌⱁ· ⱀⰰⰿ ⰱⱁⰾⰻ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰴⰻⰿⱁ ⱁⰲⱁ⸬ ⰿⱘⰶ· ⰳⱁⱄⱂⱁⰴⰰⱃⱐ· ⰱⰵⱃⰵ ⱅⰲⱁⱙ ⰲⱏⰾⱀⱘ· ⰴⰰ ⰱⱏⰹ ⰻⰿⱑⰾ ⰴⰾⱐⰰ ⱄⰵⰱⰵ ⱅⰵⱂⰾⱁ ⱂⰰⰾⱐⱅⱁ։ Ⰰ ⱁⰲⱌⰰ ⰹⰵⱄⱅ ⰱⰵⰸ ⰲⱏⰾⱀⱏⰹ։⸪
Ⱆⱄⰾⱏⰹⱎⰰⰲⱎⰻ ⱅⱁ· ⱁⰲⱌⰰ ⰻⰸⰱⱑⰳⰾⰰ ⰲ ⱃⱉⰲⱀⰻⱀⱘ։ | Ⱁⱇⰻⱌⰻⰰⰾⱀⱏⰹ ⱀⱁⱋⱀⱏⰹ ⱁⰴⰹⰵⰸⰴ ⰴⱁ Ⰽⱏⰹⰺⰵⰲⰰ։ Ⱅⱐⰿⰰ· ⰻ ⰽⱁⱃⰵⱀⱐⰵ ⱃⰵⰲⰵⱀⱐⰰ ⱂⱁⱍⱔⰾⰻ ⱃⱉⱄⱅⰵⱀⱐⰺⰵ· ⰰ ⱄⰾⱆⰳⰻ ⱂⱁⰲⱑⰴⱑⰾⰻ ⰽⱃⱉⰾⱐⱆ ⱁ ⰲⰵⱄⰵⰾⱐⱓ։.`;

  const SANITY_2 = `\
Ⱀⰰ ⰲⱏⰸⰲⱏⰹⱎⰵⱀⱁⱄⱅⰻ ⱁⰲⱌⰰ· ⰽⱅⱁⱃⰰ ⱀⰵ ⰻⱞⱑⰾⰰ ⰲⱏⰾⱀⱘ· ⱆⰲⰻⰴⱑⰾⰰ ⰽⱁⱀⱐⰵⰲ։ Ⱂⱐⱃⰲⱏⰹ ⱅⱔⰳⰰⰾ ⱅⱔⰶⰽⱏⰹ ⰲⱁⰸ· ⰲⱅⱁⱃⱏⰹ ⱀⱁⱄⰻⰾ ⰲⰵⰾⰻⰽⱁ ⰱⱃⱑⱞⱔ· ⱅⱃⰵⱅⰺⰻ ⰱⱏⱃⰸⱁ ⰲⱁⰸⰻⰾ ⱞⱘⰶⰰ։
Ⱁⰲⱌⰰ ⱃⱑⰽⰾⰰ ⰽⱁⱀⱐⰰⱞ⸬ ⸪Ⰱⱁⰾⰻ ⱞⱀⱑ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰼⱘ· ⰽⰰⰽⱁ ⱍⰾⱁⰲⱑⰽ ⰲⰾⰰⰴⰰⰹⰵ ⰽⱁⱀⱐⰰⱞⰻ։⸪
Ⰽⱁⱀⱐⰻ ⱃⱑⰽⰾⰻ⸬ ⸪Ⱄⰾⱆⱎⰰⰹ· ⱁⰲⱌⱁ· ⱀⰰⱞ ⰱⱁⰾⰻ ⱄⱐⱃⰴⱌⰵ· ⰽⱏⰳⰴⰰ ⰲⰻⰴⰻⱞⱁ ⱁⰲⱁ⸬ ⱞⱘⰶ· ⰳⱁⱄⱂⱁⰴⰰⱃⱐ· ⰱⰵⱃⰵ ⱅⰲⱁⱙ ⰲⱏⰾⱀⱘ· ⰴⰰ ⰱⱏⰹ ⰻⱞⱑⰾ ⰴⰾⱐⰰ ⱄⰵⰱⰵ ⱅⰵⱂⰾⱁ ⱂⰰⰾⱐⱅⱁ։ Ⰰ ⱁⰲⱌⰰ ⰺⰵⱄⱅ ⰱⰵⰸ ⰲⱏⰾⱀⱏⰹ։⸪
Ⱆⱄⰾⱏⰹⱎⰰⰲⱎⰻ ⱅⱁ· ⱁⰲⱌⰰ ⰻⰸⰱⱑⰳⰾⰰ ⰲ ⱃⱉⰲⱀⰻⱀⱘ։ | Ⱁⱇⰻⱌⰻⰰⰾⱀⱏⰹ ⱀⱁⱍⱐⱀⱏⰹ ⱁⰴⰹⰵⰸⰴ ⰴⱁ Ⰽⱏⰹⰺⰵⰲⰰ։ Ⱅⱐⱞⰰ· ⰻ ⰽⱁⱃⰵⱀⱐⰵ ⱃⰵⰲⰵⱀⱐⰰ ⱂⱁⱍⱔⰾⰻ ⱃⱉⱄⱅⰵⱀⱐⰺⰵ· ⰰ ⱄⰾⱆⰳⰻ ⱂⱁⰲⱑⰴⱑⰾⰻ ⰽⱃⱉⰾⱐⱆ ⱁ ⰲⰵⱄⰵⰾⱐⱓ։.`;

  test('sanity check', () => {
    expect(glag2latn(SANITY_1)).toMatchSnapshot();
  });

  test('multiple variants -> same result', () => {
    const latn1 = glag2latn(SANITY_1);
    const latn2 = glag2latn(SANITY_2);
    expect(latn1).toBe(latn2);
  });
});
