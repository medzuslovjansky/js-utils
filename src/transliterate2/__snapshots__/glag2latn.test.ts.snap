// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glag → Latn sanity check 1`] = `
"Na vȯzvyšenosti ovca, ktora ne iměla vȯlnų, uviděla końev. Pŕvy tęgal tęžky voz, vtory nosil veliko brěmę, tretji brzo vozil mųža.
Ovca rěkla końam: "Boli mně sėrdce, kȯgda viđų, kako člověk vladaje końami."
Końi rěkli: "Slušaj, ovco, nam boli sėrdce, kȯgda vidimo ovo: mųž, gospodaŕ, bere tvojų vȯlnų, da by iměl dľa sebe teplo paľto. A ovca jest bez vȯlny."
Uslyšavši to, ovca izběgla v råvninų. | Oficialny noćny odjezd do Kyjeva. Ťma, i koreńe reveńa počęli råsteńje, a slugi pověděli kråľu o veseľju.."
`;
